### PR TITLE
update importBlacklistRule to use regular expressions

### PR DESCRIPTION
**Pull request by jbsingh**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

The current importBlacklistRule only allows exact match blacklisting.  To allow greater flexibility, would like to update this rule to use regular expression pattern match blacklisting.  Existing users of the importBlacklistRule would need to make a small update to retain expected behavior, specifically they'd need to place ^ at the start and $ at the end of their existing blacklist entries as shown in the updated documentation.

#### CHANGELOG.md entry:

[rule-change] `import-blacklist` updated to use regular expression patterns for blacklisting. Place ^ at the start and $ at the end of existing blacklist entries to retain old behavior.

### CODE REVIEW

File: src/rules/importBlacklistRule.ts [hunk 1]
```diff
@@ -24,25 +24,32 @@ export class Rule extends Lint.Rules.AbstractRule {
...
-            Instead only sub modules may be imported from that module.`,
+            Disallows importing the specified modules directly via \`import\` and \`require\`.`,
         rationale: Lint.Utils.dedent`
-            Some libraries allow importing their submodules instead of the entire module.
-            This is good practise as it avoids loading unused modules.`,
-        optionsDescription: "A list of blacklisted modules.",
+            Some projects may wish to exclude imports matching certain patterns.
+
+            Examples:
+            ^loadsh$ Blacklist any import of the entire lodash module.
```
Comment (mikehaas763) [inline src/rules/importBlacklistRule.ts]: typo in the example for lodash

Comment (jbsingh) [reply src/rules/importBlacklistRule.ts]: Thanks Mike, I updated the example.

**Comment by mikehaas763** [comment]
just noticed the label, maybe there is a way to do this in a backward compatible manner?

**Comment by jbsingh** [comment]
Thanks Mike.  When originally coding this, I was hoping to make it backward compatible to eliminate impacts to any existing configuration.  Ultimately, I favored the approach chosen for the following reasons:

1. exact matching can be accomplished by regular expressions so it felt cleaner to use a single more powerful configuration option
2. supporting both existing exact match configuration in a non-breaking fashion as well as regular expression configuration would introduce some complexity in the code and configuration
3) the upgrade path is straightforward (anyone using the rule just needs to add anchors ^$ to retain existing functionality, which would be stated in the release notes that users should consult when upgrading)

If there’s strong community sentiment toward introducing the regular expression capability without a breaking change, I can definitely update the code in this PR.  Anyone have any thoughts about this?

**Comment by adidahiya** [comment]
I'd prefer to do this in a backwards compatible manner. [Regex search is slower than `indexOf`](https://jsperf.com/regexp-vs-indexof) and I think the extra complexity of supporting both configuration types is worth it here.

**Comment by jbsingh** [comment]
I made changes so that the solution is backwards compatible, but have not pushed the updates. I wanted to first find out what the protocol is. Since the changes were previously approved, should I wait for additional feedback or proceed with the backwards compatible solution?

**Comment by krawetko** [comment]
merge?

**Comment by ericanderson** [comment]
I'm sorry this took us so long to get to. It seems to be a bit at odds with import-blacklist as it exists now. Can you update it?

**Comment by JoshuaKGoldberg** [comment]
Tangentially, this should also fix #2806

**Comment by jbsingh** [comment]
I should be able to push an update within the next day or two.

**Comment by leachiM2k** [comment]
Hello guys, thank you for merging this PR. This is exactly what we've needed.
But I have a strange issue with this PR. I can't find the implemented option (and docs) in the current 5.12.1 release on npm.

**Comment by adidahiya** [comment]
@leachiM2k it wasn't released with 5.12.1. You can find out if a commit is in a release by clicking on the hash in the merge message:

image

and then looking at the tags list here:

image

there are no tags, so it is unreleased. this will be released in the next minor version.

**Pull request merged**